pkg/index/index.go

```diff
@@ -0,0 +1,223 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package index
+
+import (
+	"strings"
+	"sync"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+)
+
+// Index implements a mapping from logical cluster to (shard) URL.
+type Index interface {
+	LookupShardAndCluster(path logicalcluster.Name) (string, logicalcluster.Name, bool)
+	LookupURL(logicalCluster logicalcluster.Name) (string, bool)
+}
+
+// PathRewriter can rewrite a logical cluster path before the actual mapping through
+// the index data.
+type PathRewriter func(segments []string) []string
+
+func New(rewriters []PathRewriter) *State {
+	return &State{
+		rewriters: rewriters,
+
+		logicalClusterShards: map[logicalcluster.Name]string{},
+		shardWorkspaces:      map[string]map[logicalcluster.Name]map[string]logicalcluster.Name{},
+		shardBaseURLs:        map[string]string{},
+	}
+}
+
+// State watches ClusterWorkspaceShards on the root shard, and then starts informers
+// for every ClusterWorkspaceShard, watching the ClusterWorkspaces on them. It then
+// updates the workspace index, which maps logical clusters to shard URLs.
+type State struct {
+	rewriters []PathRewriter
+
+	lock                 sync.RWMutex
+	logicalClusterShards map[logicalcluster.Name]string                                    // logical cluster -> shard name
```

Comment:
I think that in the future, it will be `ID -> shard name`

Reply:
yes. We have to rethink the types. We need a real logical cluster and we need a workspace path.